### PR TITLE
Added initial interfaces and some sensors implementing them

### PR DESCRIPTION
**Pull request by stefan-andritoiu**

### CODE REVIEW

File: include/interfaces/iHallEffect.hpp [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+namespace upm
+{
+/**
+ * @brief Interface for Hall Effect Sensors
+*/
+  class iHallEffect
+  {
+  public:
+    virtual ~iHallEffect() {}
+    virtual bool magnetDetected() = 0;
```
Comment (pylbert) [inline include/interfaces/iHallEffect.hpp]: Do we want to encompass linear hall effect sensors?

File: examples/c++/iLight-sample.cxx [hunk 2]
```diff
@@ -0,0 +1,31 @@
+#include <iostream>
+#include <list>
+
+#include "apds9002.hpp"
+#include "bh1750.hpp"
+#include "max44009.hpp"
+
+using namespace std;
```
[the remark this answers is not in the document]
Comment (stefan-andritoiu) [reply examples/c++/iLight-sample.cxx]: Done.

File: include/interfaces/iClock.hpp [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  {
+  public:
+    virtual ~iClock() {}
+
+    /**
+     * Loads all the time values
+     *
+     * @return True if time data loaded successfully
+     */
+    virtual bool loadTime() = 0;
```
[the remark this answers is not in the document]
Comment (stefan-andritoiu) [reply include/interfaces/iClock.hpp]: Currently, there's only one sensor using it: ds1307. This sensor does not have a getTime type method, it stores the time and date in public members.

Proposal: Add new methods to the interface and to the ds1307 sensor. Example:
string getTime() //returns time as string HH:MM:SS
string getDate() //returns date DD:MM:YYYY

// and possibly others
// the return type for these can be int
string getDayOfMonth()
string getYear()

Or even use the time_t type, and have conversion methods in the interface
time_t geTime()

File: include/interfaces/iDistance.hpp [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+{
+  enum class DistanceUnit { CM, INCH };
+/**
+* @brief Interface for Distance Measuring Sensors
+*/
+  class iDistance
+  {
+  public:
+    virtual ~iDistance() {}
+    virtual int getDistance() = 0;
```
[the remark this answers is not in the document]
Comment (stefan-andritoiu) [reply include/interfaces/iDistance.hpp]: done

File: include/interfaces/iCollision.hpp [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+namespace upm
+{
+/**
+* @brief Interface for Collision Sensors
+*/
+    class iCollision
+    {
+    public:
+        virtual ~iCollision() {}
+        virtual bool isColliding() = 0;
```
[the remark this answers is not in the document]
Comment (stefan-andritoiu) [reply include/interfaces/iCollision.hpp]: done

File: include/interfaces/iEmg.hpp [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+    /**
+     * Calibrates the EMG reader
+     */
+    virtual void calibrate() = 0;
+
+    /**
+     * Measures muscle signals from the reader
+     *
+     * @return Muscle output as analog voltage
```
[the remark this answers is not in the document]
Comment (stefan-andritoiu) [reply include/interfaces/iEmg.hpp]: Currently, there's only one sensor using it: emg (and it's groove twin: groveemg). 
The C++ implementation of the emg sensor has only one function **value()**  , which is the value that is returned by **mraa_aio_read()** . 
Which according to the C implementation is the Normalized value (0.0 -> 1.0) of the voltage.

Throughout all descriptions and schematics of the Groove EMG sensor, it does not give a unit for what the sensor registers. It only says it returns the voltage, but does not specify the unit, or if this is the voltage inside the muscle, or inside the sensor, or if the two are the one and the same.

The C implementation has as "principal" function **emg_get_normalized()** which just returns 
**mraa_aio_read_float()**

But the C implementation also has:
**emg_get_raw_volts** and
**emg_get_volts**
which return the normalized value multiplied by some scales and constants.

This seem to indicate that the returned value can be converted to return a value in Volts.

All functions read off the same **mraa_aio_context** .
The C++ implementation does not include the C implementation, for these functions to also be included in C++.

I suggest adding this getVolts() function to the C++ class and the interface, and making this the standard in the interface. Hoping that **emg_get_volts** actually returns Volts (going off the name of the function), as I haven't be able to confirm that.

File: include/interfaces/iGps.hpp [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+     * @param enable true to enable the device, false otherwise.
+     */
+    virtual void enable(bool enable) = 0;
+
+    /**
+     * Set the baudrate of the device.
+     *
+     * @param baudrate The baud rate to set for the device.
+     */
+    virtual void setBaudrate(unsigned int baudrate) = 0;
```
[the remark this answers is not in the document]
Comment (stefan-andritoiu) [reply include/interfaces/iGps.hpp]: setBaudrate will be removed

File: include/interfaces/iHallEffect.hpp (hunk 1 above)
Comment (stefan-andritoiu) [reply include/interfaces/iHallEffect.hpp]: By linear hall effect sensor I'm guessing you mean one in which the output voltage linearly depends on magnetic flux density, meaning the output will be a value in a range.

I don't know how you would group those together?
We could have the magnetDetected() method return type be int. And have the range for linear sensors be (0,1).
And for the threshold sensors the output would be just be 0 or 1 (the MIN and MAX of the range).

Maybe also change the name to something like hallEffect() or magnet().

Does this sound ok?
Will this be confusing for a developer if he is using both types of sensors in his code?

Comment (Propanu) [reply include/interfaces/iHallEffect.hpp]: I'm fine with iHallEffect being a boolean interface for now.

File: include/interfaces/iEmg.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (Propanu) [reply include/interfaces/iEmg.hpp]: Volts seems adequate, and the sensor documentation/json can clarify whether it's a normalized or absolute value.

File: include/interfaces/iDistance.hpp [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+  class iDistance
+  {
+  public:
+    virtual ~iDistance() {}
+
+    /**
+     * Get distance from sensor
+     * 
+     * @return Returns the distance
+     */
```
Comment (pylbert) [inline include/interfaces/iDistance.hpp]: Units?

File: include/interfaces/iDistanceInterrupter.hpp [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+namespace upm
+{
+/**
+ * @brief Interface for Distance Interrupter Sensors
+*/
+  class iDistanceInterrupter
+  {
+  public:
+    virtual ~iDistanceInterrupter() {}
+    virtual bool objectDetected() = 0;
```
Comment (pylbert) [inline include/interfaces/iDistanceInterrupter.hpp]: Documentation needed.

File: include/interfaces/iHeartRate.hpp [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+  {
+  public:
+    virtual ~iHeartRate() {}
+
+    /**
+     * Retrieve the computed heart rate
+     *
+     * @return Computed heart rate
+     */
+    virtual int getHeartRate() = 0;
```
Comment (pylbert) [inline include/interfaces/iHeartRate.hpp]: beats per minute

File: include/interfaces/iHumidity.hpp [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+  {
+  public:
+    virtual ~iHumidity() {}
+
+    /**
+     * Measures humidity around the sensor
+     * 
+     * @return Gets the humidity value
+     */
+    virtual float getHumidity() = 0;
```
Comment (pylbert) [inline include/interfaces/iHumidity.hpp]: Relative or absolute?   Units in %

File: include/interfaces/iLight.hpp [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+*/
+    class iLight
+    {
+    public:
+        virtual ~iLight() {}
+
+        /**
+         * Get luminance
+         * 
+         * @return Luminance value
```
Comment (pylbert) [inline include/interfaces/iLight.hpp]: Units?

File: include/interfaces/iPH.hpp [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+    virtual ~iPH() {}
+
+    /**
+     * Take a number of samples and return the detected pH value.  The
+     * default number of samples is 15.
+     *
+     * @param samples The number of samples to average over, default 15
+     * @return The pH value detected
+     */
+    virtual float pH(unsigned int samples = 15) = 0;
```
Comment (pylbert) [inline include/interfaces/iPH.hpp]: Average here is not needed and doesn't match the other methods.

File: include/interfaces/iPressure.hpp [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+  class iPressure
+  {
+  public:
+    virtual ~iPressure() {}
+
+    /**
+     * Measures applied pressure
+     * 
+     * @return gets pressure value
+     */
```
Comment (pylbert) [inline include/interfaces/iPressure.hpp]: Units?

File: include/interfaces/iRotaryAngle.hpp [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+  class iRotaryAngle
+  {
+  public:
+    virtual ~iRotaryAngle() {}
+
+    /**
+     * Get rotation value from sensor data.
+     *
+     * @return rotation value.
+     */
```
Comment (pylbert) [inline include/interfaces/iRotaryAngle.hpp]: Degrees, radians?

File: include/interfaces/iTemperature.hpp [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+  public:
+
+    virtual ~iTemperature() {}
+
+    /**
+     * Measures temperature
+     * 
+     * @return temperature value
+     */
+    virtual float getTemperature() = 0;
```
Comment (pylbert) [inline include/interfaces/iTemperature.hpp]: Units?

File: include/interfaces/iVDiv.hpp [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+*/
+  class iVDiv
+  {
+  public:
+    virtual ~iVDiv() {}
+
+    /**
+     * Gets the conversion value from the sensor
+     *
+     * @return ADC conversion value
```
Comment (pylbert) [inline include/interfaces/iVDiv.hpp]: Counts?

Anything returning counts should also return a min/max or some type of range.

File: include/interfaces/iClock.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (stefan-andritoiu) [reply include/interfaces/iClock.hpp]: done

File: include/interfaces/iEmg.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (stefan-andritoiu) [reply include/interfaces/iEmg.hpp]: done

File: include/interfaces/iGps.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (stefan-andritoiu) [reply include/interfaces/iGps.hpp]: done

File: include/interfaces/iPH.hpp (hunk 13 above)
Comment (stefan-andritoiu) [reply include/interfaces/iPH.hpp]: done

File: include/interfaces/iPressure.hpp (hunk 14 above)
Comment (stefan-andritoiu) [reply include/interfaces/iPressure.hpp]: done.
Used the SI unit Pa (pascal) as the return unit for all sensors. Converted those that return bar to Pa.

**Comment by Propanu** [comment]
Builds seem OK on this. @pylbert or @stefan-andritoiu any other changes required?

**Comment by Propanu** [comment]
Merged as these are functional and most of the initial concerns have been addressed.

**Pull request closed without merging**